tasks/f3/f3_task.go

```diff
@@ -0,0 +1,285 @@
+	"github.com/filecoin-project/go-f3/manifest"
+
+	"github.com/filecoin-project/curio/deps"
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/harmony/resources"
+	"github.com/filecoin-project/curio/harmony/taskhelp"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+)
+
+const (
+	// ParticipationCheckProgressMaxAttempts defines the maximum number of failed attempts
+	// before we abandon the current lease and restart the participation process.
+	//
+	// The default backoff takes 12 attempts to reach a maximum delay of 1 minute.
+	// Allowing for 13 failures results in approximately 2 minutes of backoff since
+	// the lease was granted. Given a lease validity of up to 5 instances, this means
+	// we would give up on checking the lease during its mid-validity period;
+	// typically when we would try to renew the participation ticket. Hence, the value
+	// to 13.
+	ParticipationCheckProgressMaxAttempts = 13
+
+	// ParticipationLeaseTerm is the number of instances the miner will attempt to lease from nodes.
+	ParticipationLeaseTerm = 5
+)
+
+var log = logging.Logger("cf3")
+
+type F3ParticipationAPI interface {
+	F3GetOrRenewParticipationTicket(ctx context.Context, minerID address.Address, previous api.F3ParticipationTicket, instances uint64) (api.F3ParticipationTicket, error) //perm:sign
+	F3Participate(ctx context.Context, ticket api.F3ParticipationTicket) (api.F3ParticipationLease, error)
+	F3GetProgress(ctx context.Context) (gpbft.Instant, error)
+	F3GetManifest(ctx context.Context) (*manifest.Manifest, error)
+}
+
+type F3Task struct {
+	db  *harmonydb.DB
+	api F3ParticipationAPI
+
+	leaseTerm uint64
+
+	actors map[dtypes.MinerAddress]bool
+}
+
+func NewF3Task(db *harmonydb.DB, api F3ParticipationAPI, actors map[dtypes.MinerAddress]bool) *F3Task {
+	return &F3Task{
+		db:        db,
+		api:       api,
+		leaseTerm: ParticipationLeaseTerm,
+
+		actors: actors,
+	}
+}
+
+func (f *F3Task) Do(taskID harmonytask.TaskID, stillOwned func() bool) (done bool, err error) {
+	ctx := context.Background()
+
+	var spID int64
+	err = f.db.QueryRow(ctx, "SELECT sp_id FROM f3_tasks WHERE task_id = $1", taskID).Scan(&spID)
+	if err != nil {
+		return false, xerrors.Errorf("failed to get sp_id: %w", err)
+	}
+
+	maddr, err := address.NewIDAddress(uint64(spID))
+	if err != nil {
+		return false, xerrors.Errorf("failed to parse miner address: %w", err)
+	}
+
+	for {
+		if !stillOwned() {
+			return false, nil
+		}
+
+		var previousTicket []byte
+		err = f.db.QueryRow(ctx, "SELECT previous_ticket FROM f3_tasks WHERE task_id = $1", taskID).Scan(&previousTicket)
+		if err != nil {
+			return false, xerrors.Errorf("failed to get previous ticket: %w", err)
+		}
+
+		// Ensure that calls are made on the same node (the first call will determine the node)
+		ctx := deps.OnSingleNode(ctx)
```

Comment:
Don't we want to re-use this between loop iterations? Note: `awaitLeaseExpiry` waits till we're half-way through the lease. The idea is that the next lease will overlap with the previous one to avoid gaps.

However, that only works if we keep hitting the same node. If we don't, `tryGetF3ParticipationTicket` will keep failing and re-trying. Which is fine... I guess. It would just be _nicer_ to try the same node over and over.

Reply:
This is tricky, I may add another layer here with softer preference between the loops - OnSingleNode will fail calls if the selected backing node becomes dead iirc

Reply:
Ah, never mind, it's actually going to handle retry correctly